README.md
# sensor-tools-scada-client
Daemon to send data to sensor-tools proprietary scada service



Клиент  может находится в двух режимах:
1) live-режим.  Каждое MQTT-сообщение отправляется на сервер.
2) Archive-режим. Клиент пытается отправить архивные данные на сервер.

Условие перехода live=>Archive: переполнение очереди отправки текущих сообщений.
После переполнения очереди режим устанавливается в Archive, очередь отправки очищается.

В режиме Archive клиент запрашивает архивные данные по каналам. И пытается отправить их по-одному.


Ещё подробнее см. в коде




debian/changelog
wb-scada-client (2.0.7) stable; urgency=medium

  * fix default config

 -- Evgeny Boger <boger@contactless.ru>  Wed, 09 Nov 2016 22:31:32 +0300

wb-scada-client (2.0.6) stable; urgency=medium

  * fix cyryllic handling in archive mode

 -- Evgeny Boger <boger@contactless.ru>  Tue, 24 May 2016 22:30:18 +0300

wb-scada-client (2.0.5) stable; urgency=medium

  * properly handle Cyrillic topics

 -- Evgeny Boger <boger@contactless.ru>  Tue, 24 May 2016 12:59:02 +0300

wb-scada-client (2.0.4) stable; urgency=medium

  * fail early on invalid auth credentials

 -- Evgeny Boger <boger@contactless.ru>  Wed, 27 Apr 2016 11:45:43 +0300

wb-scada-client (2.0.3) stable; urgency=medium

  * handle errors in post_data channel (fix regression)

 -- Evgeny Boger <boger@contactless.ru>  Sun, 24 Apr 2016 14:28:44 +0300

wb-scada-client (2.0.2) stable; urgency=medium

  * properly handle server errors on cmd feed

 -- Evgeny Boger <boger@contactless.ru>  Wed, 20 Apr 2016 22:52:11 +0300

wb-scada-client (2.0.1) stable; urgency=medium

  * WIP: send meta types

 -- Evgeny Boger <boger@contactless.ru>  Fri, 08 Apr 2016 15:01:19 +0300

wb-scada-client (2.0) stable; urgency=medium

  * alpha release of renamed and reworked client 

 -- Evgeny Boger <boger@contactless.ru>  Wed, 02 Mar 2016 00:29:43 +0300

sensor-tools-scada-client (1.1) stable; urgency=medium

  * Add debug and syslog logging

 -- Evgeny Boger <boger@contactless.ru>  Tue, 8 Jul 2015 02:19:02 +0300

sensor-tools-scada-client (1.0) stable; urgency=medium

  * Initial release. (Closes: #XXXXXX)

 -- Evgeny Boger <boger@contactless.ru>  Tue, 30 Jun 2015 02:19:02 +0300


test_server.py
#!/usr/bin/env python

import socket


TCP_IP = '127.0.0.1'
TCP_PORT = 7115
BUFFER_SIZE = 20  # Normally 1024, but we want fast response

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)

s.bind((TCP_IP, TCP_PORT))
s.listen(1)
while 1:
    try:
        conn, addr = s.accept()
        print 'Connection address:', addr
        fd = conn.makefile()
        while 1:
            data = fd.readline()
            print "received data:", data
            fd.write("@OK\r\n")  # echo
            fd.flush()
    except socket.error:
        print "socker error"
    finally:
        conn.close()


test_http_server.py
#!/usr/bin/python
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
import cgi

PORT_NUMBER = 7020

# This class will handles any incoming request from
# the browser


class HttpHandler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    # Handler for the GET requests
    def do_GET(self):
        print self.headers
        print self.client_address
        self.send_response(200)
        self.send_header('Content-type', 'text/html')
        message = "Hello World !"
        self.send_header("Content-length", str(len(message)))
        self.end_headers()
        # Send the html message
        self.wfile.write(message)
        return

    # Handler for the GET requests
    def do_POST(self):
        print self.headers
        print self.client_address

        length = int(self.headers.getheader('content-length'))
        payload = self.rfile.read(length)
        print "got payload", payload

        self.send_response(200)
        self.send_header('Content-type', 'text/html')
        message = "Hello POST World !"
        self.send_header("Content-length", str(len(message)))
        self.end_headers()
        # Send the html message
        self.wfile.write(message)
        return


server = HTTPServer(('', PORT_NUMBER), HttpHandler)
print 'Started httpserver on port ', PORT_NUMBER

# Wait forever for incoming htto requests
server.serve_forever()

server.socket.close()


utils.py
import os
import re
import subprocess

try:
    from shlex import quote as sh_quote
except ImportError:
    from pipes import quote as sh_quote


def ensure_config_var(obj, var):
    if var not in obj:
        raise RuntimeError('Missing mandatory option %s in config' % var)


def get_cpuinfo_serial():
    data = open('/proc/cpuinfo').read()
    matches = re.findall('^Serial\s+: ([0-9a-f]+)$', data, re.M)
    if len(matches) > 0:
        return matches[0]
    return None


def get_stored_serial():
    fname = '/var/lib/wirenboard/serial.conf'
    if os.path.exists(fname):
        return open(fname).read().strip()
    else:
        raise RuntimeError("cannot get serial")


def get_wb_version():
    return os.environ.get('WB_VERSION')

def get_iface_ip(iface):
    cmd = subprocess.Popen("/sbin/ifconfig %s | grep \"inet addr\" | awk -F: '{print $2}' | awk '{print $1}'" % sh_quote(iface), shell=True, stdout=subprocess.PIPE)
    ip = cmd.stdout.readline().strip()
    return ip or None


mqttclient.py
#coding: utf8
import sys, struct
try:
    import mosquitto
    TMQTTClient = mosquitto.Mosquitto
except ImportError:
    import paho.mqtt.client as mosquitto
    from paho.mqtt.client import MQTT_ERR_NO_CONN

    class TPahoMosquittoUnicodeFix(mosquitto.Mosquitto):
        def _pack_str16(self, packet, data):
            """ this version doesn't try to reencode the strings"""
            if sys.version_info[0] < 3:
                if isinstance(data, bytearray):
                    packet.extend(struct.pack("!H", len(data)))
                    packet.extend(data)
                elif isinstance(data, str):
                    udata = data
                    pack_format = "!H" + str(len(udata)) + "s"
                    packet.extend(struct.pack(pack_format, len(udata), udata))
                elif isinstance(data, unicode):
                    udata = data.encode('utf-8')
                    pack_format = "!H" + str(len(udata)) + "s"
                    packet.extend(struct.pack(pack_format, len(udata), udata))
                else:
                    raise TypeError
            else:
                if isinstance(data, bytearray) or isinstance(data, bytes):
                    packet.extend(struct.pack("!H", len(data)))
                    packet.extend(data)
                elif isinstance(data, str):
                    udata = data.encode('utf-8')
                    pack_format = "!H" + str(len(udata)) + "s"
                    packet.extend(struct.pack(pack_format, len(udata), udata))
                else:
                    raise TypeError

        def subscribe(self, topic, qos=0):
            topic_qos_list = None
            if isinstance(topic, str):
                if qos<0 or qos>2:
                    raise ValueError('Invalid QoS level.')
                if topic is None or len(topic) == 0:
                    raise ValueError('Invalid topic.')
                topic_qos_list = [(topic, qos)]

            if isinstance(topic, unicode):
                if qos<0 or qos>2:
                    raise ValueError('Invalid QoS level.')
                if topic is None or len(topic) == 0:
                    raise ValueError('Invalid topic.')
                topic_qos_list = [(topic.encode('utf8'), qos)]

            elif isinstance(topic, tuple):
                if topic[1]<0 or topic[1]>2:
                    raise ValueError('Invalid QoS level.')
                if topic[0] is None or len(topic[0]) == 0 or not isinstance(topic[0], str):
                    raise ValueError('Invalid topic.')
                topic_qos_list = [(topic[0].encode('utf-8'), topic[1])]
            elif isinstance(topic, list):
                topic_qos_list = []
                for t in topic:
                    if t[1]<0 or t[1]>2:
                        raise ValueError('Invalid QoS level.')
                    if t[0] is None or len(t[0]) == 0 or not isinstance(t[0], str):
                        raise ValueError('Invalid topic.')
                    topic_qos_list.append((t[0].encode('utf-8'), t[1]))

            if topic_qos_list is None:
                raise ValueError("No topic specified, or incorrect topic type.")

            if self._sock is None and self._ssl is None:
                return (MQTT_ERR_NO_CONN, None)

            return self._send_subscribe(False, topic_qos_list)


    TMQTTClient = TPahoMosquittoUnicodeFix

if __name__ == '__main__':
    m = TMQTTClient()
    m.connect('localhost')
    m.subscribe('sdfsdf')
    m.subscribe('sdfsdf')
    m.subscribe('sdfsdf')
    m.subscribe(u'ываыва')
    m.subscribe(u'ываыва')
    m.subscribe('ываыва')
    m.subscribe('ываыва')
    m.publish('ываыва','ываыва')


wb-scada-client.py
#!/usr/bin/python
# coding: utf-8
import json
import argparse
import Queue
import datetime
import time
import os
import collections

from mqttclient import TMQTTClient, mosquitto

import logging
import logging.handlers

from mqttrpc.client import TMQTTRPCClient
# from jsonrpc.exceptions import JSONRPCError

from backends import TServerConnectionBase, TRequestException
import sensor_tools_backend
import wbscada_http_backend
import utils

scada_backends = {
    'sensor-tools': sensor_tools_backend,
    'wbscada-http': wbscada_http_backend
}


logger = logging.getLogger('')
handler = logging.handlers.SysLogHandler(address='/dev/log')
formatter = logging.Formatter('wb-scada-client: %(levelname)s:%(name)s:%(message)s')
handler.setFormatter(formatter)
logger.addHandler(handler)

ch = logging.StreamHandler()
ch_formatter = logging.Formatter('%(asctime)s - %(name)s:%(levelname)s:%(message)s')
ch.setFormatter(ch_formatter)
logger.addHandler(ch)

logging.getLogger('').setLevel(logging.WARNING)



class TSCADAClient(object):
    DEFAULT_QUEUE_SIZE = 100
    DEFAULT_RPC_TIMEOUT = 10
    RPC_MAX_ROWS_PER_QUERY = 100

    SERVER_FAIL_RETRY_TIMEOUT = 10
    SERVER_FAIL_RETRY_ATTEMPTS = 5

    RPC_ERROR_RETRY_TIMEOUT = 10

    def load_last_success_item(self):
        if os.path.exists(self.last_success_item_fname):
            try:
                parts = open(self.last_success_item_fname).read().strip().split()
                self.last_success_timestamp = float(parts[0])
                self.last_success_uid = int(parts[1])
                return
            except:
                logging.warning("can't load last sucessfull item data from file")

    def set_last_success_item(self, timestamp, uid=None):
        self.last_success_timestamp = timestamp
        if uid is not None:
            self.last_success_uid = uid

        if not os.path.exists(os.path.dirname(self.last_success_item_fname)):
            os.makedirs(os.path.dirname(self.last_success_item_fname))

        open(self.last_success_item_fname, 'wt').write("%f %d\n" %
                                                       (self.last_success_timestamp,
                                                        self.last_success_uid))

    def __init__(self, config):
        self.mqtt_client = TMQTTClient()

        self.config = config

        self.live_mode = False
        self.live_queue = Queue.Queue(
            maxsize=int(self.config.get('queue_size', self.DEFAULT_QUEUE_SIZE)))

        self.last_success_timestamp = 0
        self.last_success_uid = -1

        self.data_dir = self.config.get('data_dir', "/var/lib/wb-scada-client")

        self.last_success_item_fname = os.path.join(self.data_dir, 'last_success_item.dat')
        self.load_last_success_item()

        utils.ensure_config_var(self.config, 'connection_backend')

        if self.config['connection_backend'] not in scada_backends:
            raise RuntimeError("Unknown connection backend %s" % self.config['connection_backend'])

        backend = scada_backends[self.config['connection_backend']]

        self.scada_conn = backend.ServerConnection(
            self.config,
            mqtt_client=self.mqtt_client,
            data_dir=self.data_dir,
            parent = self,
        )

        if self.config.get('mqtt_username'):
            self.mqtt_client.username_pw_set(
                self.config['mqtt_username'], self.config.get('mqtt_password', ''))

        self.rpc_client = TMQTTRPCClient(self.mqtt_client)

        self.rpc_timeout = float(self.config.get('rpc_timeout', self.DEFAULT_RPC_TIMEOUT))
        self.mqtt_client.on_message = self.on_mqtt_message
        self.mqtt_client.on_connect = lambda mosq, obj, rc: self.on_mqtt_connect(mosq, obj, rc)

        self.mqtt_client.connect(
            self.config.get('mqtt_host', 'localhost'), self.config.get('mqtt_port', 1883))
        self.mqtt_client.loop_start()

        self.channels_meta = collections.defaultdict(dict)

    def on_mqtt_message(self, mosq, obj, msg):
        logging.debug("got mqtt message on topic %s" % msg.topic)
        if self.rpc_client.on_mqtt_message(mosq, obj, msg):
            return

        if not mosquitto.topic_matches_sub('/devices/+/controls/#', msg.topic):
            return

        parts = msg.topic.split('/')
        device_id = parts[2].decode('utf8')
        control_id = parts[4].decode('utf8')

        channel = (device_id, control_id)

        if mosquitto.topic_matches_sub('/devices/+/controls/+/meta/+', msg.topic):
            meta = parts[6]
            self.channels_meta[channel][meta] = msg.payload
            logging.debug("got meta %s %s %s " % (channel, meta, msg.topic))

        elif mosquitto.topic_matches_sub('/devices/+/controls/+', msg.topic):
            if not self.live_mode:
                return

            try:
                self.live_queue.put_nowait((datetime.datetime.now(),
                                            channel, msg.payload))
            except Queue.Full:
                logging.info("Setting live_mode to false")
                self.live_mode = False

                # do not call Queue methods inside 'with' block!
                # Queue.join() won't work after clearing in this way
                with self.live_queue.mutex:
                    self.live_queue.queue.clear()

    def on_mqtt_connect(self, mosq, obj, rc):
        for device_id, control_id in self.scada_conn.get_channels():
            topic = u"/devices/%s/controls/%s" % (device_id, control_id)
            topic = topic.encode('utf8')
            self.mqtt_client.subscribe(topic + "/meta/+")
            self.mqtt_client.subscribe(topic)

    def process_data_item(self, channel, dt, value, live_mode):

        # пытаемся отправить строку на сервер. Если словили timeout, пытаемся ещё,
        #  пока находимся в live mode
        # если получили FAIL, то (пока) делаем то же самое, но без реконнекта

        server_fail_number = 0

        while True:
            if live_mode:
                if not self.live_mode:
                    return False

            try:
                resp = self.scada_conn.do_request(dt, live=live_mode,
                                                      channels={channel: value})
            except TRequestException:
                logging.warning("error while performing request, reconnecting")
                self.scada_conn.reconnect()
            else:
                if resp:
                    return True
                else:
                    server_fail_number += 1

                    if server_fail_number >= self.SERVER_FAIL_RETRY_ATTEMPTS:
                        logging.warning("server wasn't able to process request %d times, giving up" %
                                        server_fail_number)
                        return False
                    else:
                        logging.warning("server can't process request, trying again")
                        time.sleep(self.SERVER_FAIL_RETRY_TIMEOUT)

    def do_archive_mode_work(self):
        """ return True if there is no more data to send"""

        resp = self.rpc_client.call('db_logger', 'history', 'get_values',
                                    {
                                        'channels': self.scada_conn.get_channels(),
                                        'timestamp': {
                                            'gt': self.last_success_timestamp,
                                        },

                                        'uid': {
                                            'gt': self.last_success_uid,
                                        },
                                        'limit': self.RPC_MAX_ROWS_PER_QUERY,
                                    }, self.rpc_timeout)

        for row in resp.get(u'values', ()):
            channel = (row[u'device'], row[u'control'])

            value = row[u'value']
            dt = datetime.datetime.fromtimestamp(row[u'timestamp'])

            if self.process_data_item(channel, dt, value, live_mode=False):
                self.set_last_success_item(row[u'timestamp'], row[u'uid'])

        if not resp.get(u'has_more'):
            return True

    def loop(self):
        """ main loop"""

        self.scada_conn.reconnect()

        while 1:
            if self.live_mode:
                dt, channel, value = self.live_queue.get(timeout=1E100)
                if self.process_data_item(channel, dt, value, live_mode=True):
                    self.set_last_success_item(time.mktime(dt.timetuple()))

            else:
                logging.debug("not in live mode")
                try:
                    done = self.do_archive_mode_work()
                except AssertionError:
                    raise
                except:
                    logging.exception("error in archive mode handler")
                    time.sleep(self.RPC_ERROR_RETRY_TIMEOUT)
                else:
                    if done:
                        logging.debug("going back to live mode")
                        self.live_mode = True


def main():
    parser = argparse.ArgumentParser(description='WB SCADA client', add_help=False)

    parser.add_argument('config_file', type=str,
                        help='Config file location')

    parser.add_argument('-d', '--debug', dest='debug', action='store_true',
                        help='debug')

    args = parser.parse_args()
    config = json.load(open(args.config_file))

    if args.debug or config.get('debug', None):
        logging.getLogger().setLevel(logging.DEBUG)

    scada_client = TSCADAClient(config)
    scada_client.loop()


if __name__ == "__main__":
    main()


sensor_tools_backend.py
import Queue
import socket
import math
import logging
import time
import utils
from backends import TServerConnectionBase, TRequestException


class TSensorToolsServerConnection(TServerConnectionBase):
    MAX_RECONNECT_INTERVAL = 60
    MIN_RECONNECT_INTERVAL = 0.1
    RESPONSE_OK = '@OK'
    RESPONSE_FAIL = '@FAIL'
    POLLING_INTERVAL = 60
    DEFAULT_SAVING_INTERVAL = 300
    DEFAULT_TCP_TIMEOUT = 20

    def __init__(self, config):
        self.config = config

        utils.ensure_config_var(self.config, 'server_host')
        utils.ensure_config_var(self.config, 'server_port')
        utils.ensure_config_var(self.config, 'channels')

        self.host = self.config['server_host']
        self.port = self.config['server_port']
        self.timeout = self.config.get('tcp_timeout', self.DEFAULT_TCP_TIMEOUT)
        self.saving_interval = self.config.get('saving_interval_prop',
                                               self.DEFAULT_SAVING_INTERVAL)

        self.client_id = self.config['client_id']

        self.channel_map = {}
        for channel_id, channel in self.config['channels'].iteritems():
            device_id, control_id = channel
            self.channel_map[(str(device_id), str(control_id))] = int(channel_id)

        self.sock = None
        self.fd = None

    def get_channels(self):
        return [(device_id, control_id) for device_id, control_id in self.channel_map.iterkeys()]

    def reconnect(self):
        if self.sock:
            try:
                self.sock.close()
            except:
                pass

        reconnect_interval = self.MIN_RECONNECT_INTERVAL
        while True:
            try:
                self.connect()
            except socket.error, err:
                logging.exception("connect failed")
                time.sleep(reconnect_interval)
                reconnect_interval = min(self.MAX_RECONNECT_INTERVAL, reconnect_interval * 2)
            else:
                return

    def connect(self):
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.connect((self.host, self.port))
        self.sock.settimeout(self.timeout)
        self.fd = self.sock.makefile()

    def _escape(self, var):
        if isinstance(var, unicode):
            var = var.encode('utf-8', 'ignore')
        if not isinstance(var, str):
            var = str(var)

        return var.replace(':', '').replace(';', '').replace('#', '')

    def prepare_request(self, timestamp, live, channels):
        parts = [("id", self._escape(self.client_id)),
                 ('datetime', timestamp.strftime('%Y%m%d%H%M%S')),
                 ('saving_interval', self.saving_interval),
                 ('live', int(live)),
                 ]

        for k, v in channels.iteritems():
            try:
                _v = float(v)
            except ValueError:
                pass
            else:
                if not math.isnan(_v) and not math.isinf(_v):
                    channel_id = self.channel_map[k]
                    if channel_id is not None:
                        parts.append(('t%d' % channel_id, v))

        var_data = ";".join("%s:%s" % (k, self._escape(v)) for k, v in parts)

        return "".join(("#", var_data, "#"))

    def do_request(self, req):
        """ sends request and returns response (without trailing newline).
            throws TRequestException in case socket is closed or on socket timeout"""

        logging.debug("Sending request: '%s'" % req)

        try:
            self.fd.write(req)
            self.fd.write('\n')
            self.fd.flush()

            response = self.fd.readline()
            if response:
                # strip \r\n
                response = response[:-2]
            else:
                response = None

            if response == self.RESPONSE_OK:
                return True
            elif response == self.RESPONSE_FAIL:
                return False
            elif response is None:
                logging.warning("server closed the connection")
                raise TRequestException()
            else:
                logging.warning("unexpected answer from server: '%s'" % response)
                raise TRequestException()

        except (socket.timeout, socket.error):
            logging.warning("request timeout")
            raise TRequestException()


ServerConnection = TSensorToolsServerConnection


sensor-tools-client.py
#!/usr/bin/python
#coding: utf-8
import json
import argparse
import Queue
import socket
import datetime
import time
import os
import mosquitto
import logging
import logging.handlers
import math

from mqttrpc.client import TMQTTRPCClient
from jsonrpc.exceptions import JSONRPCError



logger = logging.getLogger('')
handler = logging.handlers.SysLogHandler(address = '/dev/log')
formatter = logging.Formatter('sensor-tools-client: %(levelname)s:%(name)s:%(message)s')
handler.setFormatter(formatter)
logger.addHandler(handler)

ch = logging.StreamHandler()
ch_formatter = logging.Formatter('%(asctime)s - %(name)s:%(levelname)s:%(message)s')
ch.setFormatter(ch_formatter)
logger.addHandler(ch)

logging.getLogger('').setLevel(logging.WARNING)

class TServerConnection(object):
    MAX_RECONNECT_INTERVAL = 60
    MIN_RECONNECT_INTERVAL = 0.1
    RESPONSE_OK = '@OK'
    RESPONSE_FAIL = '@FAIL'
    POLLING_INTERVAL = 60
    DEFAULT_SAVING_INTERVAL = 300

    def __init__(self, host, port, timeout=1, saving_interval = None):
        self.host = host
        self.port = port
        self.timeout = timeout
        self.sock = None
        self.fd = None

        self.saving_interval = saving_interval or self.DEFAULT_SAVING_INTERVAL


    def reconnect(self):
        if self.sock:
            try:
                self.sock.close()
            except:
                pass

        reconnect_interval = self.MIN_RECONNECT_INTERVAL
        while True:
            try:
                self.connect()
            except socket.error, err:
                logging.exception("connect failed")
                time.sleep(reconnect_interval)
                reconnect_interval = min(self.MAX_RECONNECT_INTERVAL, reconnect_interval * 2)
            else:
                return



    def connect(self):
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.connect((self.host, self.port))
        self.sock.settimeout(self.timeout)
        self.fd = self.sock.makefile()

    def _escape(self, var):
        if  isinstance(var, unicode):
            var = var.encode('utf-8', 'ignore')
        if not isinstance(var, str):
            var = str(var)


        return var.replace(':', '').replace(';','').replace('#','')

    def format_request(self, client_id, timestamp, live, channels):
        parts = [ ("id", self._escape(client_id)),
                  ('datetime', timestamp.strftime('%Y%m%d%H%M%S')),
                  ('saving_interval' , self.saving_interval),
                  ('live' , int(live)),
                ]

        for k, v in channels.iteritems():
            try:
                _v = float(v)
            except ValueError:
                pass
            else:
                if not math.isnan(_v) and not math.isinf(_v):
                    parts.append(('t%d' % k, v))


        var_data = ";".join( "%s:%s" % (k, self._escape(v)) for k, v in parts)

        return "".join( ("#", var_data, "#"))

    def do_request(self, req):
        """ sends request and returns response (without trailing newline).
            if socket is closed, returns None
            throws socket.timeout on timeout """

        logging.debug("Sending request: '%s'" % req)

        self.fd.write(req)
        self.fd.write('\n')
        self.fd.flush()

        response = self.fd.readline()
        if response:
            # strip \r\n
            return response[:-2]
        else:
            return None


class TSensorToolsClient(object):
    DEFAULT_TCP_TIMEOUT = 20
    DEFAULT_QUEUE_SIZE = 100
    DEFAULT_RPC_TIMEOUT = 10
    RPC_MAX_ROWS_PER_QUERY = 100

    SERVER_FAIL_RETRY_TIMEOUT = 10
    SERVER_FAIL_RETRY_ATTEMPTS = 5

    RPC_ERROR_RETRY_TIMEOUT = 10

    def _ensure_config_var(self, obj, var):
        if var not in obj:
            raise RuntimeError('Missing mandatory option %s in config' % var)





    def load_last_success_item(self):
        if os.path.exists(self.last_success_item_fname):
            try:
                parts = open(self.last_success_item_fname).read().strip().split()
                self.last_success_timestamp = float(parts[0])
                self.last_success_uid  = int(parts[1])
                return
            except:
                logging.warning("can't load last sucessfull item data from file")


    def set_last_success_item(self, timestamp, uid = None):
        self.last_success_timestamp = timestamp
        if uid is not None:
            self.last_success_uid = uid

        if not os.path.exists(os.path.dirname(self.last_success_item_fname)):
            os.makedirs(os.path.dirname(self.last_success_item_fname))

        open(self.last_success_item_fname, 'wt').write("%f %d\n" % (self.last_success_timestamp, self.last_success_uid))


    def __init__(self, config):
        self.mqtt_client = mosquitto.Mosquitto()


        self.config = config

        self.live_mode = False
        self.live_queue = Queue.Queue(maxsize=int(self.config.get('queue_size', self.DEFAULT_QUEUE_SIZE)))

        self._ensure_config_var(self.config, 'server_host')
        self._ensure_config_var(self.config, 'server_port')

        self._ensure_config_var(self.config, 'channels')

        self.scada_client_id = self.config['client_id']

        self.last_success_timestamp = 0
        self.last_success_uid  = -1
        self.last_success_item_fname = self.config.get('last_success_file', '/var/lib/sensor-tools/last_success_item.dat')
        self.load_last_success_item()


        self.channel_map = {}
        for channel_id, channel in self.config['channels'].iteritems():
            device_id, control_id = channel
            self.channel_map[(str(device_id), str(control_id))]  = int(channel_id)


        self.scada_conn = TServerConnection(self.config['server_host'], self.config['server_port'],
                                            self.config.get('tcp_timeout', self.DEFAULT_TCP_TIMEOUT),
                                            saving_interval = self.config.get('saving_interval_prop'))


        if self.config.get('mqtt_username'):
            self.mqtt_client.username_pw_set(self.config['mqtt_username'], self.config.get('mqtt_password', ''))


        self.rpc_client = TMQTTRPCClient(self.mqtt_client)

        self.rpc_timeout = float(self.config.get('rpc_timeout', self.DEFAULT_RPC_TIMEOUT))
        self.mqtt_client.on_message = self.on_mqtt_message
        self.mqtt_client.on_connect = self.on_mqtt_connect

        self.mqtt_client.connect(self.config.get('mqtt_host', 'localhost'), self.config.get('mqtt_port', 1883))
        self.mqtt_client.loop_start()

    def on_mqtt_message(self, mosq, obj, msg):
        if self.rpc_client.on_mqtt_message(mosq, obj, msg):
            return


        if not self.live_mode:
            return

        if not mosquitto.topic_matches_sub('/devices/+/controls/+', msg.topic):
            return

        parts = msg.topic.split('/')
        device_id = parts[2]
        control_id = parts[4]

        channel_id  = self.channel_map.get((device_id, control_id))
        if channel_id:
            try:
                self.live_queue.put_nowait((datetime.datetime.now(), channel_id, msg.payload))
            except Queue.Full, exc:
                logging.info("Setting live_mode to false")
                self.live_mode = False

                # do not call Queue methods inside 'with' block!
                # Queue.join() won't work after clearing in this way
                with self.live_queue.mutex:
                    self.live_queue.queue.clear()



    def on_mqtt_connect(self, mosq, obj, rc):
        for device_id, control_id in self.channel_map.iterkeys():
            topic = "/devices/%s/controls/%s" % (device_id, control_id)
            self.mqtt_client.subscribe(topic)

    def process_data_item(self, channel_id, dt, value, live_mode):
        req = self.scada_conn.format_request(self.scada_client_id, dt, live=live_mode, channels={channel_id : value})

        # пытаемся отправить строку на сервер. Если словили timeout, пытаемся ещё, пока находимся в live mode
        # если получили FAIL, то (пока) делаем то же самое, но без реконнекта

        server_fail_number = 0

        while True:
            if live_mode:
                if not self.live_mode:
                    return False

            try:
                resp = self.scada_conn.do_request(req)
            except (socket.timeout, socket.error):
                logging.warning("request timeout, reconnecting")
                self.scada_conn.reconnect()
            else:
                if resp == self.scada_conn.RESPONSE_OK:
                    return True
                elif resp == self.scada_conn.RESPONSE_FAIL:
                    server_fail_number += 1

                    if server_fail_number >= self.SERVER_FAIL_RETRY_ATTEMPTS:
                        logging.warning("server returned @FAIL %d times, giving up" % server_fail_number)
                        return False
                    else:
                        logging.warning("server returned @FAIL, trying again")
                        time.sleep(self.SERVER_FAIL_RETRY_TIMEOUT)
                elif resp is None:
                    logging.warning("server closed the connection")
                    self.scada_conn.reconnect()
                else:
                    logging.warning("unexpected answer from server: '%s'" % resp)
                    self.scada_conn.reconnect()


    def do_archive_mode_work(self):
        """ return True if there is no more data to send"""

        resp =  self.rpc_client.call('db_logger', 'history', 'get_values', {
                                        'channels': self.channel_map.keys(),
                                        'timestamp' : {
                                            'gt': self.last_success_timestamp,
                                        },

                                        'uid' : {
                                            'gt': self.last_success_uid,
                                        },
                                        'limit' : self.RPC_MAX_ROWS_PER_QUERY,
                                    }, self.rpc_timeout)


        for row in resp.get(u'values', ()):
                channel_id = self.channel_map.get((str(row[u'device']),str(row[u'control'])))

                assert channel_id is not None

                value = str(row[u'value'])
                dt = datetime.datetime.fromtimestamp(row[u'timestamp'])

                if self.process_data_item(channel_id, dt, value, live_mode=False):
                    self.set_last_success_item(row[u'timestamp'], row[u'uid'])

        if not resp.get(u'has_more'):
            return True


    def loop(self):
        """ main loop"""

        self.scada_conn.reconnect()

        while 1:
            if self.live_mode:
                dt, channel_id, value  = self.live_queue.get(timeout=1E100)
                if self.process_data_item(channel_id, dt, value, live_mode=True):
                    self.set_last_success_item(time.mktime(dt.timetuple()))

            else:
                logging.debug("not in live mode")
                try:
                    done = self.do_archive_mode_work()
                except AssertionError:
                    raise
                except:
                    logging.exception("error in archive mode handler")
                    time.sleep(self.RPC_ERROR_RETRY_TIMEOUT)
                else:
                    if done:
                        logging.debug("going back to live mode")
                        self.live_mode = True





def main():
    parser = argparse.ArgumentParser(description='Sensor-Tools TCP SCADA client', add_help=False)

    parser.add_argument('config_file', type=str,
                     help='Config file location')

    parser.add_argument('-d', '--debug', dest='debug', action='store_true',
                     help='debug')

    args = parser.parse_args()
    config = json.load(open(args.config_file))


    if args.debug or config.get('debug', None):
        logging.getLogger().setLevel(logging.DEBUG)


    scada_client = TSensorToolsClient(config)
    scada_client.loop()


if __name__ == "__main__":
    main()
